src/cron/calculateTopReaders.ts

```diff
@@ -0,0 +1,119 @@
+import type { Cron } from './cron';
+
+import { logger } from '../logger';
+import { readFile } from 'fs/promises';
+import path from 'path';
+import { UserTopReader } from '../entity';
+import { endOfMonth, subMonths } from 'date-fns';
+
+type TopReaderQueryResult = {
+  keyword: string;
+  keywordRank: number;
+  userId: string;
+  userViewCount: number;
+  rank: number;
+};
+
+const LIMIT_PER_KEYWORD = 10;
+const MINIMUM_UNIQUE_VIEWERS = 500;
+const TOP_READERS_TO_SELECT = 100;
+
+export const calculateTopReaders: Cron = {
+  name: 'calculate-top-readers',
+  handler: async (con) => {
+    logger.info('calculateTopReaders');
+    // Set issuedAt to be the last day of the previous month
+    const issuedAt = new Date(
+      endOfMonth(subMonths(new Date(), 1)).setHours(0, 0, 0, 0),
+    );
+
+    const topReaders: Record<string, string[]> = {};
+    const userIds = new Map<string, boolean>();
+    const keywords: Record<string, Omit<TopReaderQueryResult, 'keyword'>[]> =
+      {};
+
+    try {
+      const sql = await readFile(
+        path.join(__dirname, 'calculateTopReaders.sql'),
+        'utf8',
+      );
+
+      const result = await con.query<TopReaderQueryResult[]>(sql, [
```

Comment:
run it on read replica?

Reply:
can use `queryReadReplica`

Reply:
Ideally, yes. but read replica has issue with long queries, as it'll terminate them.
We can adjust this setting, but as we get more users and views, the query duration will likely increase, and we'll need to adjust this

So I would prefer running it on master

Reply:
ok, let's leave it on master for now, but in the future would like it to be on rr